x86: not all fields of promoted struct need regs

CarolEidt opened the pull request:
This fixes DevDiv bug 278375, and includes a test case. It also improves the code generation for the general case of promoted structs, though tuning is still needed.
The bug was a case where, after decomposing the long fields, we had 7 entries in the GT_FIELD_LIST, each of which was requesting a register.
Since Lowering won't know which fields may be in registers, it must be prepared to handle fields in memory. Also, if the fields are not register candidates, we should make them contained and push them directly.
In most cases, we should use push for promoted structs (they are generally small-ish, and pushes of lclVars and constants are cheaper than loading into a register).
This could be optimized for consecutive float or double fields.

CarolEidt commented:
@dotnet/jit-contrib PTAL

CarolEidt commented:
@dotnet/jit-contrib PTAL

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -7424,20 +7427,42 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+// - This argument contains any GC pointers OR
+// - It is a GT_FIELD_LIST OR
+// - It is less than 16 bytes in size.
+#ifdef DEBUG
```
You could avoid clang-format messing with the indenting here by using CLANG_FORMAT_COMMENT_ANCHOR;

BruceForstall commented on src/jit/codegenxarch.cpp:
```diff
@@ -7424,20 +7427,42 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+        default:
+            assert(!"Uninitialized GenTreePutArgStk::Kind");
+            break;
+            ;
```
extraneous semicolon?

CarolEidt replied:
That's what I get for running jit-format -f and not looking at the results! I've fixed these (I suspect I missed the extra semicolon before because I'd put them both on the same line)

---
In reply to: [86234819](https://github.com/dotnet/coreclr/pull/7847#discussion_r86234819) [](ancestors = 86234819)

BruceForstall commented on src/jit/lowerxarch.cpp:
```diff
@@ -2104,20 +2105,48 @@ void Lowering::TreeNodeInfoInitPutArgStk(GenTreePutArgStk* putArgStk)
-            // TODO-X86-CQ: we could probably improve codegen here by marking all of the operands to field nodes that
-            // we are going to `push` on the stack as reg-optional.
+            // For x86 we must mark all integral fields as contained or reg-optional, and handle them
+            // accordingly in code generation, since we may have up to 8 fields, which cannot all be in
```
>// accordingly in code generation, since we may have up to 8 fields, which cannot all be in [](start = 12, length = 91)

Why "up to 8 fields"? Is that obvious for some reason?

CarolEidt replied:
It's because `GT_FIELD_LIST` is used for promoted structs, which can have up to 4 fields. If they are all longs, that will make 8 entries in the list.

BruceForstall commented:
>     }

Do you expect this case to ever occur? That is, can this instead be noway_assert ?

---
Refers to: src/jit/codegenxarch.cpp:7690 in 585d2ad. [](commit_id = 585d2ade6c25dbc204d657e613071fbed45c2473, deletion_comment = False)

BruceForstall commented:
:shipit:

CarolEidt commented on src/jit/lowerxarch.cpp:
```diff
@@ -2129,11 +2158,35 @@ void Lowering::TreeNodeInfoInitPutArgStk(GenTreePutArgStk* putArgStk)
+            putArgStk->gtPutArgStkKind = GenTreePutArgStk::Kind::Unroll;
         }
+#endif
         return;
```
No, thanks for catching that. I was originally thinking that we needed to support the `Unroll` case as well, but that adds non-trivial complexity and I think the `Push` approach is reasonable to use for most cases of promoted structs.

---
In reply to: [86237042](https://github.com/dotnet/coreclr/pull/7847#discussion_r86237042) [](ancestors = 86237042)

CarolEidt commented:
>     }

No, it shouldn't occur. This was pre-existing (though the code has all moved around a bit), but I will change it.

---
In reply to: [257994531](https://github.com/dotnet/coreclr/pull/7847#issuecomment-257994531) [](ancestors = 257994531)

---
Refers to: src/jit/codegenxarch.cpp:7690 in 6dd4c90. [](commit_id = 6dd4c90a83d6fbafcedd526874124aa4a7a655ec, deletion_comment = False)

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+            // assigned a register, and which is therefore contained.
+            // Unlike genConsumeReg(), it handles the case where no registers are being consumed.
+            genConsumeRegs(fieldNode);
+            regNumber argReg = fieldNode->gtRegNum;
```
This is not valid if `fieldNode` is reg-optional. This should instead be `regNumber argReg = fieldNode->isContainedSpillTemp() ? REG_NA : fieldNode->gtRegNum;`.

pgavlin replied:
Also, is there a reason to do this here rather than prior to the stack adjustment? If we did this earlier, we could fold the `if (!m_pushStkArg)` code below into the `else` block of the `if (fieldIsSlot)` statement above.

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
         }
-
-        // Adjust the stack if necessary.
-        if (currentOffset != 0)
```
Why was this code removed? If this situation is not thought to be possible, we should have an assert to verify that it does not occur.

CarolEidt replied:
I will add an assert. I don't believe there can ever be padding at the beginning of a struct.

---
In reply to: [86258424](https://github.com/dotnet/coreclr/pull/7847#discussion_r86258424) [](ancestors = 86258424)

pgavlin replied:
Yes, I think you're right: this was handling padding at the end of the struct in the case that we had pre-adjusted the stack, which I now notice we're no longer doing.

CarolEidt replied:
Right - but after your comment it occurred to me that there could possibly be padding at the beginning of a struct if `ExplicitLayout` is used. So I think we need to just retain that code.

---
In reply to: [86262042](https://github.com/dotnet/coreclr/pull/7847#discussion_r86262042) [](ancestors = 86262042)

pgavlin commented:
Can you provide an example of the generated code?

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+                    }
+                    if (needByteReg)
+                    {
+                        noway_assert((genRegMask(tmpReg) & RBM_BYTE_REGS) != 0);
```
I don't think that this should be a `noway_assert`: the other reg-related asserts above are simple `assert`s, so I don't think this adds much value. Furthermore, I maintain that we should not use `noway_assert`s unless we can prove that they are in fact recoverable.

CarolEidt replied:
OK I will change to an assert.

---
In reply to: [86260756](https://github.com/dotnet/coreclr/pull/7847#discussion_r86260756) [](ancestors = 86260756)

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+                adjustment -= pushSize;
+                while (adjustment != 0)
+                {
+                    inst_IV(INS_push, 0);
```
Is this series of pushes smaller than a `sub esp`?

CarolEidt replied:
Hmmm ... it is possible that a sub esp,4 (the common case) would be the same size (I think it would also be 2 bytes). But there will almost never be more than one of these, because it would only be possible with explicit layout where you happen to have two slots of padding. jit32 uses push eax which is a single byte, but I'm not sure whether it needs to ensure that eax doesn't contain, e.g. a GC ref.

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+                fieldType         = genActualType(fieldType);
+                unsigned pushSize = genTypeSize(fieldType);
+                assert((pushSize % 4) == 0);
+                noway_assert((adjustment % TARGET_POINTER_SIZE) == 0);
```
This had better be guaranteed by `roundUp`... I would leave this as an `assert` (or remove it entirely).

CarolEidt replied:
I'll remove it.

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7488,115 +7514,183 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
             GenTree* const fieldNode   = current->Current();
             const unsigned fieldOffset = current->gtFieldOffset;
             var_types      fieldType   = current->gtFieldType;
+            const int      fieldSize   = genTypeSize(fieldType);
```
It doesn't look like this is used any longer.

CarolEidt replied:
I removed it.

CarolEidt commented:
@pgavlin 
> Can you provide an example of the generated code?

Here is a repro case. This is a struct with 2 long fields, a double and a ref. The longs get decomposed into 4 int fields:

```
N275 (  3,  2) [000085] -------N----       t85 =    lclFld    int    V17 tmp11        u:3[+0] REG NA
N277 (???,???) [000503] ------------      t503 =    lclFld    int    V17 tmp11        [+4] (last use) REG NA
N279 (  3, 10) [000369] ------------      t369 =    const     int    2 REG NA
N281 (???,???) [000506] ------------      t506 =    const     int    0 REG NA
N283 (  3,  4) [000371] ------------      t371 =    dconst    double 3.0000000000000000 REG mm0 $101
N285 (  1,  1) [000373] ------------      t373 =    lclVar    ref    V20 tmp14        u:3 eax (last use) REG eax RV $303
                                                 /--*  t373   ref    
                                                 +--*  t371   double 
                                                 +--*  t506   int    
                                                 +--*  t369   int    
                                                 +--*  t503   int    
                                                 +--*  t85    int    
N287 (  1,  1) [000374] ------------      t374 = *  <fldList> void   ref at offset 24 REG NA $19d
```
Here's the code that's generated:
```
IN0049: 000129 movsd    xmm0, qword ptr [@RWD00]
IN004a: 000131 push     eax
IN004b: 000132 sub      esp, 8
IN004c: 000135 movsd    qword ptr [esp], xmm0
IN004d: 00013A push     0
IN004e: 00013C push     2
IN004f: 00013E push     dword ptr [V17+0x4 ebp-48H]
IN0050: 000141 push     dword ptr [V17 ebp-4CH]
```

CarolEidt commented:
test Windows_NT x86 ryujit jitstressregs2

CarolEidt commented:
test Windows_NT x86 ryujit jitstressregs3

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit jitstressregs2

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit jitstressregs3

CarolEidt commented:
@pgavlin I think I addressed all your feedback.
Thanks for catching the spill temp and padding issues!

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit jitstressregs3

sivarv commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

CarolEidt commented:
@dotnet-bot test this please

CarolEidt commented:
@dotnet-bot test this please

CarolEidt commented:
@dotnet-bot retest this please

CarolEidt commented:
@dotnet-bot test OSX x64 Checked Build and Test

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2_jitstressregs8
@dotnet-bot test Windows_NT x86 ryujit Checked jitstress2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

CarolEidt commented:
@dotnet-bot test Linux ARM Emulator Cross Debug Build

CarolEidt commented:
@dotnet/jit-contrib PTAL
I've addressed the issues exposed by the stress tests (and in the course of stress testing I discovered a test that was throwing an exception but not reporting failure, so I've included a fix for that test).
The issue (DevDiv 278375) is that we can have up to 8 fields in a GT_FIELD_LIST, so not all of them can have registers. The short summary of what I’ve done is:
* In the case of lclVars that are NOT register candidates, I make them contained because they can be pushed without being loaded into a register.
  * Doing this alone would actually have fixed the bug that was filed for this, DevDiv 278375, because the int fields of a “promoted” long are not currently register candidates. However, I wasn’t comfortable with that as even a short-term fix.
* In the case of integer constants, I made them contained also because they can also be directly pushed.
  * Note that this doesn’t actually apply to sub-word constants, but for that see below.
* All other non-floating-point fields are marked “RegOptional” at use.
  * If these are pointer-sized, they can be directly pushed from the stack (either their home location for lclVars or the spill slot for others).
* However, if there are any sub-pointer-sized fields, we can’t push them directly, since the stack must remain pointer-sized aligned, so I reserve an internal register. If any of them are bytes, it is marked as needing to be a byteable register.
* The current “RegOptional” implementation won’t spill anything that’s being used at the current location. I changed this to allow such an interval to be spilled if it is marked “RegOptional”. That’s necessary to allow the above to work in the case where there are too many nodes that are register candidates (even if optional).

However, there are a number of things that are non-optimal:
* (Not related to this fix) We currently push the fields in reverse order, which is what we want to do for pointer-sized fields. However, for smaller fields we should keep them in the list order so that we can push the first field in the slot as 32-bits, and then store the remaining fields.
* We don’t require the byte fields to be in byteable registers, but it would be nice if we could preference them that way.
* If we have sub-pointer-sized fields, we always allocate an internal register, even though we may not need it. One strategy (which I think might best be done as a follow-on work item):
  * If any of the sub-pointer-sized fields are *either* tree temps or last-use lclVars:
    * For the first such field, require a register (i.e. don’t make it RegOptional). This register can then be also used as a temporary register to load up the remaining sub-pointer-sized fields.
    * For the first byte field, make it require a register – this may be the same as above, or in addition to. One could fine-tune this but it’s probably not really worth it.

sivarv commented on src/jit/lsra.cpp:
```diff
@@ -5724,13 +5724,14 @@ regNumber LinearScan::allocateBusyReg(Interval* current, RefPosition* refPositio
         // because we'll always find something with a later nextLocation, but it can happen in stress when
         // we have LSRA_SELECT_NEAREST.
-        if ((nextLocation == refLocation) && !refPosition->isFixedRegRef)
+        if ((nextLocation == refLocation) && !refPosition->isFixedRegRef && nextRefPosition->RequiresRegister())
```
This might have some good diffs on amd64 as well. You might want to check whether any of cqperf benchmark methods impacted.

CarolEidt replied:
Over all the tests, frameworks & benchmarks there is a very small (48 byte) code size increase.
For the benchmarks the net is a very small overall (218) byte code size reduction:

Improvements:
Platform.Graphics.Typography.TrueType.ScanConverter:CalcSpline() goes from  3205 to  2953, diff of 252 (7.86%)
IDEAEncryption:cipher_idea() goes from  1135 to  1124, diff of 11 ( 0.97%)

Regressions:
Platform.Graphics.Typography.TrueType.LocalGraphicState:IUPx() goes from  2156 to  2186, diff of 30 (1.37%)
Platform.Graphics.Typography.TrueType.ScanConverter:MeasureGlyph() goes from  2566 to  2581, diff of 15 (0.58%)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -3282,6 +3282,8 @@ void CodeGen::genStructPutArgUnroll(GenTreePutArgStk* putArgNode)
     regNumber longTmpReg = REG_NA;
 #ifdef _TARGET_X86_
+    // On x86 we use an XMM register for both 16 and 8-byte chunks, but if it's
+    // less that 16 bytes, we will just be using pushes
```
>    // less that 16 bytes [](start = 0, length = 25)

Typo "less than"

sivarv commented:
> //                    memory; false otherwise.

This comment line now can be removed since isSrcInMemory arg is removed.

---
Refers to: src/jit/codegenxarch.cpp:7430 in 8661bb9. [](commit_id = 8661bb9cd45333667457725d03d6cb7c287c29d0, deletion_comment = False)

sivarv commented on src/jit/lowerxarch.cpp:
```diff
@@ -2119,10 +2155,28 @@ void Lowering::TreeNodeInfoInitPutArgStk(GenTreePutArgStk* putArgStk)
+        // adjust the stack once for those fields. The latter is really best done in code generation, but
+        // this tuning should probably be undertaken as a whole.
+        // Also, if there are  floating point fields, it may be better to use the "Unroll" mode
+        // of copying the struct as a whole, if the fields are not register candidates.
```
Please file tracking git issue(s) for possible improvements that were uncovered in this work.

CarolEidt replied:
Will do. I thought I would wait and see the response from my comments on this latest commit.

---
In reply to: [88320293](https://github.com/dotnet/coreclr/pull/7847#discussion_r88320293) [](ancestors = 88320293)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -7478,8 +7504,8 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+    const unsigned argSize = putArgStk->getArgSize();
+    assert((argSize % TARGET_POINTER_SIZE) == 0);
 
     // TODO-Cleanup: Handle this in emitInsMov() in emitXArch.cpp?
```
Is this comment relevant now?

CarolEidt replied:
I would say not; I'll remove it.

---
In reply to: [88321469](https://github.com/dotnet/coreclr/pull/7847#discussion_r88321469) [](ancestors = 88321469)

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -7492,22 +7518,27 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+        const bool preAdjustedStack = genAdjustStackForPutArgStk(putArgStk);
+        // For now, we only support the "push" case; we will push a full slot for the first field of each slot
+        // within the struct.
+        assert((putArgStk->isPushKind()) && !preAdjustedStack && m_pushStkArg);
```
IMO, GT_PUTARG_STK(GT_FIELD_LIST) is a bit complex to follow and too big to deserve a routine of its own.

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -7649,6 +7751,45 @@ void CodeGen::genPutArgStk(GenTreePutArgStk* putArgStk)
+        {
+            ins = ins_Store(type);
+        }
+        assert(genIsValidFloatReg(srcReg));
```
Not sure I understood why this special handling of TYP_LONG.  A comment explaining the scenario where it is used will be helpful here.

CarolEidt replied:
Done. This is only used for the "unroll" case.

sivarv commented:
:shipit:

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7423,29 +7426,237 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+            m_pushStkArg = false;
+            if (adjustment != 0)
+            {
+                // This moves the stack pointer to (fieldOffset + sizeof(fieldType)) to account for the `push`.
```
I think this comment is out-of-date: the `sub` moves SP to `fieldOffset`.

CarolEidt replied:
Fixed.

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7423,29 +7426,237 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+            if ((argReg == REG_NA) || needByteReg)
+            {
+                assert(varTypeIsIntegralOrI(fieldNode));
+                if (tmpReg == REG_NA)
```
IMO we ought to move the initialization of `tmpReg` out of the loop. We could simply have something like:
```
regNumber tmpReg = REG_NA;
if (putArgStk->gtRsvdRegs != RBM_NONE)
{
    assert(genCountBits(putArgStk->gtRsvdRegs) == 1);
    tmpReg = genRegNumFromMask(putArgStk->gtRsvdRegs);
    assert(genIsValidIntReg(tmpReg));
}
```

pgavlin replied:
Note that this would also simplify the case above to `if (needByteReg)`.

pgavlin commented on src/jit/codegenxarch.cpp:
```diff
@@ -7423,29 +7426,237 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+            }
+        }
+
+        // Consume the register, if any, for this field. Note that genConsumeRegs() will appropriately
```
If we were to move the call to `genConsumeRegs` and the calculation of `argReg` before the code above that adjusts the stack, we could avoid the second check for `!m_pushStkArg` by folding the body of that `if` statement in with the body of the `else` above.

pgavlin commented:
@dotnet-bot test this please

CarolEidt commented on src/jit/codegenxarch.cpp:
```diff
@@ -7423,29 +7426,237 @@ unsigned CodeGen::getBaseVarForPutArgStk(GenTreePtr treeNode)
+            adjustment -= pushSize;
+            while (adjustment != 0)
+            {
+                inst_IV(INS_push, 0);
```
That will be captured in the perf-related issues.

---
In reply to: [88963166](https://github.com/dotnet/coreclr/pull/7847#discussion_r88963166) [](ancestors = 88963166)

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

pgavlin commented:
@dotnet-bot test CentOS7.1 x64 Debug Build and Test

CarolEidt commented:
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x10
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs0x80
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs1
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs2
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs3
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs4
@dotnet-bot test Windows_NT x86 ryujit Checked jitstressregs8

pgavlin commented:
@CarolEidt: let's go ahead and merge this and take care of any leftover feedback afterwards.

CarolEidt commented:
@pgavlin - yes, will do. It took me 3 rounds to get the handling correct for `GT_UMOD` but I believe I've got it now!

CarolEidt commented:
@pgavlin:
> take care of any leftover feedback afterwards

I opened #8289 to cover the remaining perf-related issues. We may want to consider breaking it out into separate issues, but they're probably best addressed as a whole.
